lib/mayaUsd/ufe/StagesSubject.cpp

```diff
@@ -72,10 +72,12 @@ bool inAttributeChangedNotificationGuard()
     return attributeChangedNotificationGuardCount.load() > 0;
 }
 
-// TODO: This should be an unordered_multimap to prevent notifications from
-// overwriting earlier recorded notifications for the same attribute. See
-// MAYA-110878 for more information on why this change isn't already made.
-std::unordered_map<Ufe::Path, TfToken> pendingAttributeChangedNotifications;
+// Keep an array of the pending attribute change notifications. Using a vector
+// for two main reasons:
+// 1) Order of notifs must be maintained.
+// 2) Allow notif with same path, but different token. At worst the check is linear
+//    in the size of the vector (which is the same as an unordered_multimap).
+std::vector<std::pair<Ufe::Path, TfToken>> pendingAttributeChangedNotifications;
```

Comment:
Change is pretty simple. I just changed the container to vector (for reasons give here). I ran all the UFE tests and they all passed with no test changes required. I believe this is due to the previous changes around line 420 which ignore inert prim changes.

Reply:
I think that vector will not be fast enough for the duplicate notification elimination at line 95. While the worst case is the same, the average case or unordered_map is constant while the average case the vector is linear in the size of the vector.

I think to do better you can have a second container which is a std::unordered_set< std::pair<Ufe::Path, TfToken> >. The idea is you use the set to test if the std::pair<Ufe::Path, TfToken> has already been saved in the vector in constant time, and you use the vector to maintain ordering. We trade off a little constant time & some space, but we avoid the linear search.

Reply:
I do see your point, but I wonder if its really needed. I added some tf_debug statements to print out when a notif gets delayed. I ran all the "UFE" tests and the largest size the pendingAttributeChangedNotifications vector got was only 2. Granted the tests aren't real-world data, but I think that when normally running the notif guard will expire and send the pending notifs long before they ever have a chance to build up. @ppt-adsk What do you think?

Reply:
Hard to say, I tend to think @seando-adsk you're right, but it's a code complexity vs. performance guarantee tradeoff.  Can we perhaps put up a message if the vector size grows above some threshold?

Reply:
The case I'm worried about is someone writing a script that iterates over everything in the scene and changes each one a little bit, say changing the purpose of each rprim. That is going to create N messages and have O(N^2) complexity because it is doing a linear search.

If you really don't like the unordered_set we can leave it as-is and ask QA to play around with some heavy scenes and see if they hit any performance issues.

Reply:
If you look [here](https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/ufe/UsdAttribute.cpp#L54) you can see where we use the notif guard - basically when setting a single usd attribute. In your scenario on each prim we change a little bit we will enter/leave a notif guard. So I don't ever see this notif guard growing beyond just a few notifs.

Reply:
Okay sounds good to me!